### PR TITLE
Fix: ECS Atomicity problem

### PR DESCRIPTION
**Pull request by mikhail-dcl**
## ECS Atomicity Problem

https://miro.com/app/board/uXjVMnrITv4=/?moveToWidget=3458764588435139719&cot=14

**Problem definition**

image

**Solution**

image

## Visibility Problem

https://miro.com/app/board/uXjVMnrITv4=/?moveToWidget=3458764588438892911&cot=14

image
image

## Synchronization between Global and Scene Worlds is no longer required

It happened because we locked out the whole cycle in the frame from the changes, so we guarantee no changes from SDK can squeeze in.

## JS API promises problem

- If API requires a call that can be only completed on the main thread, it switches to it (subscribes to the player loop), then the `Promise` (JS Promise) returned as a continuation from the main thread
- Then such "continuation" is marshalled via ClearScript
- And the rest of the JS logic (whatever it is) executes on the main thread
- Considering it can await changes from `EngineAPI` it leads eventually to a dead-lock as now the whole player-loop cycle is locked out from changes (and thus from completion of `EngineAPI.sendCrdt`

**Solution** 

Disconnect the promise returned to the JS side:

```
public static object ToDisconnectedPromise<T>(this UniTask<T> uniTask)
        {
            var completionSource = new UniTaskCompletionSource<T>();

            UniTask.RunOnThreadPool(async () =>
                    {
                        try
                        {
                            T result = await uniTask;
                            await UniTask.SwitchToThreadPool();
                            completionSource.TrySetResult(result);
                        }
                        catch (Exception e)
                        {
                            await UniTask.SwitchToThreadPool();
                            completionSource.TrySetException(e);
                        }
                    })
                   .Forget();

            return completionSource.Task.ToPromise();
        }
```

## How to test the changes?

1. Check scenes/worlds with interactivity troubles such as lifts and doors.
2. Check scene with visibility problems
3. There should be no regression with asset bundles and primitives loading

### CODE REVIEW

**Review by m3taphysics — APPROVED** [review]

File: Explorer/Assets/DCL/Interaction/PlayerOriginated/Utility/HoverFeedbackUtils.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ public static class HoverFeedbackUtils
     {
         public static void TryIssueLeaveHoverEventForPreviousEntity(in PlayerOriginRaycastResult raycastResult, in GlobalColliderEntityInfo previousEntityInfo)
         {
-            using (previousEntityInfo.EcsExecutor.Sync.GetScope())
+            // using (previousEntityInfo.EcsExecutor.Sync.GetScope())
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/DCL/Interaction/PlayerOriginated/Utility/HoverFeedbackUtils.cs:13]: Forgot to delete, no more synchronization required

File: Explorer/Assets/Scripts/Global/Dynamic/RealmNavigator.cs [hunk 2]
```diff
@@ -22,6 +22,7 @@
 using System.Threading;
 using DCL.UserInAppInitializationFlow;
 using ECS.Prioritization.Components;
+using System.Linq;
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/Global/Dynamic/RealmNavigator.cs:25]: yeah, I think we are fine, it's one-timer

File: Explorer/Assets/Scripts/ECS/Abstract/Events/ClearEntityEventsSystem.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        public override void Dispose()
+        {
+            foreach (EntityEventBuffer entityEventBuffer in eventBuffers)
+                entityEventBuffer.Dispose();
+        }
+
+        protected override void Update(float t)
+        {
+            foreach (EntityEventBuffer entityEventBuffer in eventBuffers)
+                entityEventBuffer.Clear();
```
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/ECS/Abstract/Events/ClearEntityEventsSystem.cs:27]: In this system on dispose:
```
public override void Dispose()
        {
            foreach (EntityEventBuffer entityEventBuffer in eventBuffers)
                entityEventBuffer.Dispose();
        }
```

[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/ECS/Abstract/Events/ClearEntityEventsSystem.cs:27]:
```
public override void Dispose()
        {
            pool.Release(this);
        }
```

File: Explorer/Assets/Scripts/ECS/StreamableLoading/Common/Systems/LoadSystemBase.cs [hunk 4]
```diff
@@ -31,20 +30,16 @@ public abstract class LoadSystemBase<TAsset, TIntention> : BaseUnityLoopSystem w
...
         private CancellationTokenSource cancellationTokenSource;
 
         private bool systemIsDisposed;
 
-        protected LoadSystemBase(World world, IStreamableCache<TAsset, TIntention> cache, MutexSync mutexSync) : base(world)
+        protected LoadSystemBase(World world, IStreamableCache<TAsset, TIntention> cache) : base(world)
         {
             this.cache = cache;
-            this.mutexSync = mutexSync;
+            // this.mutexSync = mutexSync;
```
Comment (Kinerius) [inline Explorer/Assets/Scripts/ECS/StreamableLoading/Common/Systems/LoadSystemBase.cs:42]: comment

File: Explorer/Assets/Scripts/ECS/StreamableLoading/Common/Systems/LoadSystemBase.cs [hunk 5]
```diff
@@ -160,7 +155,7 @@ private void Execute(in Entity entity, ref StreamableLoadingState state, ref TIn
             StreamableLoadingResult<TAsset>? result, AssetSource source,
             IAcquiredBudget acquiredBudget)
         {
-            using MutexSync.Scope sync = mutexSync.GetScope();
+            // using MutexSync.Scope sync = mutexSync.GetScope();
```
Comment (Kinerius) [inline Explorer/Assets/Scripts/ECS/StreamableLoading/Common/Systems/LoadSystemBase.cs:158]: comment

**Review by Kinerius — APPROVED** [review]
comment

**Pull request merged**